Mixed shard repair reproducer

Deexie opened the pull request:
Reproducer for mixed shard repair to choose the best solution for https://github.com/scylladb/scylladb/issues/18269.

Sets up a 3-node cluster on AWS with 1TB of data and runs repair.

It will be run with jenkins with the following configurations:
- on master; each node has the same number of shards (60)
- on master; nodes will have 60, 59, 58 shards
- POC1 (https://github.com/scylladb/scylladb/pull/20295); 60, 59, 58
- POC2 (https://github.com/scylladb/scylladb/pull/20296); 60, 59, 58

Deexie committed:
feature(mixed-shards): custom shard numbers

Add custom shard number config for Scylla clusters.

denesb commented:
I am not familiar with the SCT code, but the description looks good to me.
Did you get a chance to run the test? How do the numbers look?

Deexie commented:
- change instance type to i3.16xlarge

Deexie commented:
- change shards count

Deexie commented:
- change loaders instance
- split data population

Deexie commented:
master-60-59-58 
time: 936.2321102619171
argus: https://argus.scylladb.com/test_runs?state=WyI4YTI3MGEyYS05OWE2LTQwYzYtODkxNS1lMzhiOGFiOGQ2OGMiXQ
non-LSA memory: ![image](https://github.com/user-attachments/assets/1f6e58bc-ab07-46b1-a6a5-6227e7732bd3)

Deexie commented:
master 60-60-60
test duration: 1h 24min
repair time: 553.5129013061523 (9 min)
![image](https://github.com/user-attachments/assets/e586ecc9-18f4-4774-ba4d-57b8078f0e05)

Deexie commented:
poc1-60-59-58
test duration: 3h 2min
repair time: 5473.398208618164 (1.5h)
![image](https://github.com/user-attachments/assets/dbe2d1a4-9dd7-4318-97f5-30234a400a16)

Deexie commented:
poc2-60-59-58
failed after: 7h 45min
```
02:09:38  error running operation: std::system_error (error system:104, recv: Connection reset by peer)
02:09:38  ----- LAST WARNING EVENT -----------------------------------------------------
02:09:38  2024-09-05 19:38:43.928 <2024-09-05 19:38:43.699>: (DatabaseLogEvent Severity.WARNING) period_type=one-time event_id=c571cbf1-5131-4da8-8563-3d5ed86ec7bb: type=WARNING regex=(^WARNING|!\s*?WARNING).*\[shard.*\] line_number=109442 node=ubuntu-mixed-sh-db-node-a7786369-2
02:09:38  2024-09-05T19:38:43.699+00:00 ubuntu-mixed-sh-db-node-a7786369-2  !WARNING | scylla[15842]:  [shard  0: gms] seastar_memory - oversized allocation: 1069056 bytes. This is non-fatal, but could lead to latency and/or fragmentation issues. Please report: at 0x6102d3e 0x6103350 0x6103658 0x5bafde2 0x5bb2585 0x45e9158 0x45e255a 0x5c0591f 0x5c06e9a 0x5c08077 0x5c07428 0x5b97593 0x5b968f3 0x13cf2f5 0x13d0cb0 0x13cd713 /opt/scylladb/libreloc/libc.so.6+0x2a087 /opt/scylladb/libreloc/libc.so.6+0x2a14a 0x13cad94
02:09:38  ----- LAST NORMAL EVENT ------------------------------------------------------
02:09:38  2024-09-05 19:38:10.473: (PrometheusAlertManagerEvent Severity.NORMAL) period_type=end event_id=d69143fc-2503-4eb1-b248-61a7a9171077 duration=1h35m59s: alert_name=InstanceDown node=10.4.1.235 start=2024-09-05T18:02:07.408Z end=2024-09-05T18:06:07.408Z description=10.4.1.235 has been down for more than 30 seconds. updated=2024-09-05T18:02:07.412Z state=active fingerprint=45469a7e312b47e8 labels={'alertname': 'InstanceDown', 'cluster': 'my-cluster', 'dc': 'eu-west-1', 'instance': '10.4.1.235', 'job': 'scylla', 'monitor': 'scylla-monitor', 'severity': '3'}
02:09:38  ================================================================================
```

asias commented:
@Deexie How did you execute the new sct test introduced in this PR? Do you run through Jenkins? Could you share the details?

Deexie commented:
> @Deexie How did you execute the new sct test introduced in this PR? Do you run through Jenkins? Could you share the details?

Yes, I run it with Jenkins, here you have the link: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/aleksandra-martyniuk/job/mixed-shard-repair-reproducer/

Here the list of runs, whenever I mention "master" I mean an image of https://github.com/Deexie/scylla/tree/mixed-shard-repair-master built in `master-64-63-62#2`. Repair is run on the first node on the list. Configurations that contain a node with 60 shard run on `i3.16xlarge`, others on `i3en.6xlarge`

`master-64-63-62` -
- 8 -  the run on master with 60, 59, 58 shards (turned out we cannot use all shard in a machine)
- 9 - same as above but with repair logger level set to debug
- all previous runs failed due to wrong configuration

`master-64-64-64` - run on master with 60, 60, 60 shards
`poc1-64-63-62` - run on https://github.com/Deexie/scylla/tree/mixed-shard1 with 60, 59, 58 shards
`poc2-64-63-62` - run on https://github.com/Deexie/scylla/tree/mixed-shard2 with 60, 59, 58 shards
`master-60-48-32`  - master with 60, 48, 32
`master-32-48-60` (here I made versions mess)
- 2 - master 32, 48, 60 (same as above but repair is started on the smallest node)
- 3 - master 60, 2, 2 (repair logger set to debug)
- 4, 5, 6 - master 16, 16, 12

`master-16-12-12-on-all` - master 16, 16, 12, but repair is started on all nodes
`master-16-16-12-bootstrap` - master 16, 16; data is inserted; new node is bootstrapped

Deexie commented:
`master-16-16-12-bootstrap` looks promising. It failed after 2,5h due to timeouts, both semaphore and not.
On shard dumped semaphore stats twice with 9min gap:
```
Sep 20 13:21:55.843556 ubuntu-mixed-sh-db-node-341faace-2 scylla[6664]:  [shard 13:strm] reader_concurrency_semaphore - Semaphore streaming with 10/10 count and 3230160/231903068 memory resources: timed out, dumping permit diagnostics:
                                                                        permits        count        memory        table/operation/state
                                                                        10        10        3154K        keyspace1.standard1/shard-reader/active/await
                                                                        86        0        0B        keyspace1.standard1/shard-reader/waiting_for_admission
                                                                        
                                                                        96        10        3154K        total
                                                                        
                                                                        Stats:
                                                                        permit_based_evictions: 357875
                                                                        time_based_evictions: 0
                                                                        inactive_reads: 0
                                                                        total_successful_reads: 0
                                                                        total_failed_reads: 0
                                                                        total_reads_shed_due_to_overload: 0
                                                                        total_reads_killed_due_to_kill_limit: 0
                                                                        reads_admitted: 399954
                                                                        reads_enqueued_for_admission: 358156
                                                                        reads_enqueued_for_memory: 0
                                                                        reads_admitted_immediately: 41884
                                                                        reads_queued_because_ready_list: 0
                                                                        reads_queued_because_need_cpu_permits: 2079
                                                                        reads_queued_because_memory_resources: 356077
                                                                        reads_queued_because_count_resources: 0
                                                                        reads_queued_with_eviction: 21117
                                                                        total_permits: 63357
                                                                        current_permits: 96
                                                                        need_cpu_permits: 10
                                                                        awaits_permits: 10
                                                                        disk_reads: 10
                                                                        sstables_read: 43
                                                                
                                                                
Sep 20 13:30:56.120683 ubuntu-mixed-sh-db-node-341faace-2 scylla[6664]:  [shard 13:strm] reader_concurrency_semaphore - Semaphore streaming with 10/10 count and 3244683/231903068 memory resources: timed out, dumping permit diagnostics:
                                                                        permits        count        memory        table/operation/state
                                                                        1        1        24K        keyspace1.standard1/shard-reader/active/need_cpu
                                                                        1        0        9K        keyspace1.standard1/shard-reader/evicted
                                                                        53        0        0B        keyspace1.standard1/shard-reader/waiting_for_admission
                                                                        
                                                                        55        1        33K        total
                                                                        
                                                                        Stats:
                                                                        permit_based_evictions: 568285
                                                                        time_based_evictions: 0
                                                                        inactive_reads: 9
                                                                        total_successful_reads: 0
                                                                        total_failed_reads: 0
                                                                        total_reads_shed_due_to_overload: 0
                                                                        total_reads_killed_due_to_kill_limit: 0
                                                                        reads_admitted: 610364
                                                                        reads_enqueued_for_admission: 568547
                                                                        reads_enqueued_for_memory: 0
                                                                        reads_admitted_immediately: 41884
                                                                        reads_queued_because_ready_list: 0
                                                                        reads_queued_because_need_cpu_permits: 15636
                                                                        reads_queued_because_memory_resources: 552911
                                                                        reads_queued_because_count_resources: 0
                                                                        reads_queued_with_eviction: 22374
                                                                        total_permits: 63409
                                                                        current_permits: 64
                                                                        need_cpu_permits: 1
                                                                        awaits_permits: 0
                                                                        disk_reads: 10
                                                                        sstables_read: 43
```

Metrics look better than ever (in terms of expectations):
![image](https://github.com/user-attachments/assets/ea5d6654-65f8-4183-8098-4f7127333a65)
![image](https://github.com/user-attachments/assets/1f04414d-6145-4aea-890f-6d5738c428c2)


I reran the test to check if we get similar results and run the same test on https://github.com/Deexie/scylla/tree/mixed-shard-master-no-false-timeouts (version of the `master-16-16-12-bootstrap` with @denesb's false timeout fix)

Deexie commented:
Update: 
The issue was finally reproduced with a bootstrap test:
- test: test_mixed_shard_bootstrap in https://github.com/scylladb/scylla-cluster-tests/blob/f5073743607b98bb1f0127d52aeffa190e9e1529/mixed_shard_repair_test.py
- configuration: https://github.com/scylladb/scylla-cluster-tests/blob/f5073743607b98bb1f0127d52aeffa190e9e1529/test-cases/features/mixed-shard-bootstrap.yaml
- jenkins pipeline: https://github.com/scylladb/scylla-cluster-tests/blob/f5073743607b98bb1f0127d52aeffa190e9e1529/jenkins-pipelines/oss/features/features-mixed-shard-bootstrap.jenkinsfile

In mixed cluster repair lasted ~1h27m, while the same test with the same shard numbers (3x16) - ~27m30s. This was the result for master with applied @denesb's https://github.com/scylladb/scylladb/pull/20703 (without it mixed shard case was constantly failing with timeout).

roydahan commented on test-cases/features/mixed-shard-bootstrap.yaml:
```diff
@@ -0,0 +1,20 @@
+test_duration: 1200
+
+n_db_nodes: 2
```
Is it intentional to have here only 2 nodes?

The smp list is for 3 nodes and it changes the third node that doesn't exist...

Deexie replied:
Yes, the third node is added during the test and it gets the value from the list

roydahan commented on jenkins-pipelines/oss/features/features-fixed-shard-bootstrap.jenkinsfile:
I suggest putting these files (and their related test cases) under another nesting level of the "feature name".
(Although it's not really a feature and usually we don't try to cover things like that in a such way).

roydahan commented:
Generally I understand the motivation to try and cover as much as possible, but going forward in terms of regression testing, these tests aren't going to be automatically trigger.
Hence, usually we pick the interesting cases and create longevities out of them and/or adding randomness.

Deexie commented:
> Generally I understand the motivation to try and cover as much as possible, but going forward in terms of regression testing, these tests aren't going to be automatically trigger. Hence, usually we pick the interesting cases and create longevities out of them and/or adding randomness.

The test is meant to compare the possible solutions rather than to be triggered regularly. Most of the cases are to be dropped - they didn't reproduce the issue or they required modifications, I just keep them here to get back to it if we needed it (but I don't think we will). I will clean it up, once we decide that some part of it (if anything) should be merged.

Deexie commented:
- rf=3
- 4 nodes

Deexie committed:
test(mixed-shard-repair): add test for mixed shard repair

Deexie commented:
- extend custom shard num for all types of cluster
- cleanup test cases

denesb commented:
@pehala please review.

pehala reviewed (commented):
I see we have [db_nodes_shards_selection](https://github.com/scylladb/scylla-cluster-tests/blob/master/docs/configuration_options.md#user-content-db_nodes_shards_selection) option, that we use in asymetrical test cases, would this work here? Or could be extended to work here, without so many changes in the core?

Deexie commented:
> I see we have [db_nodes_shards_selection](https://github.com/scylladb/scylla-cluster-tests/blob/master/docs/configuration_options.md#user-content-db_nodes_shards_selection) option, that we use in asymetrical test cases, would this work here? Or could be extended to work here, without so many changes in the core?

With this option, we can have different numbers of shards, but they are taken randomly. Here, we can specify the exact number of shards per node. If we reuse `db_nodes_shards_selection`, then I think we still need to propagate the number of shards on each node.

 Originally, it was done for AWS only. Maybe that's a good way.

roydahan commented:
Please note that "features" kind of tests are tests that aren't being triggered regularly (especially not automatically).
It's a good way to test a specific feature quickly, but not part of regression testing of releases.

Till now we had the asymmetrical longevities to exercise this path, it's obviously not enough since it didn't detect the issue we had in the field.
I recommend to either refactor current longevity or add a new longevity that will exercise this "feature".

denesb commented:
@Deexie what is the status of this PR?

denesb commented:
@Deexie what is the status here?

Deexie commented:
> @Deexie what is the status here?

I'm getting back to it. Currently, the PR contains a feature that enables setting the shard number for each server and the test that was used in the mixed shard issue. I do not see how to achieve what's tested here without the custom shard num feature, nor how to make it a regression test that runs periodically.

> I see we have [db_nodes_shards_selection](https://github.com/scylladb/scylla-cluster-tests/blob/master/docs/configuration_options.md#user-content-db_nodes_shards_selection) option, that we use in asymetrical test cases, would this work here? Or could be extended to work here, without so many changes in the core?

@pehala please see my response above (https://github.com/scylladb/scylla-cluster-tests/pull/8435#issuecomment-2538954601). Do you think that the change may get in as is? Does it need additional testing? Should I run it with each backend and check whether the number of cores is as specified?

I don't think we can go with `db_nodes_shards_selection`.



> Please note that "features" kind of tests are tests that aren't being triggered regularly (especially not automatically). It's a good way to test a specific feature quickly, but not part of regression testing of releases.
> 
> Till now we had the asymmetrical longevities to exercise this path, it's obviously not enough since it didn't detect the issue we had in the field. I recommend to either refactor current longevity or add a new longevity that will exercise this "feature".

@roydahan This test wasn't meant to run periodically. The bug was examined based on metrics and logs. I don't know how to convert this into longevity.

roydahan commented:
> @roydahan This test wasn't meant to run periodically. The bug was examined based on metrics and logs. I don't know how to convert this into longevity.

Maybe one simple way is to change the current "asymmetric" longevities configuration to use "nodes_smp: [X, Y, Z]" instead of the current random, with number of smp that we think will stress this feature the most.
You can do that by either adding another configuration file like the one here https://github.com/scylladb/scylla-cluster-tests/blob/master/configurations/db-nodes-shards-random.yaml and set some of the longevities that uses this one with your new config file.

fruch commented on sdcm/cluster_aws.py:
```diff
@@ -893,6 +894,7 @@ def __init__(self, ec2_ami_id, ec2_subnet_id, ec2_security_group_ids,  # pylint:
                  n_nodes=3,
                  params=None,
                  node_type: str = 'scylla-db',
+                 nodes_smp=[]
```
don't initialize an argument with a mutable list
set it as None, and later in code initialize the list if needed

fruch commented on sdcm/cluster.py:
```diff
@@ -3211,7 +3214,7 @@ class BaseCluster:  # pylint: disable=too-many-instance-attributes,too-many-publ
     # pylint: disable=too-many-arguments,too-many-locals,too-many-branches
     def __init__(self, cluster_uuid=None, cluster_prefix='cluster', node_prefix='node', n_nodes=3, params=None,
-                 region_names=None, node_type=None, extra_network_interface=False, add_nodes=True):
+                 region_names=None, node_type=None, extra_network_interface=False, add_nodes=True, nodes_smp=[]):
```
don't initialize an argument with a mutable list
set it as None, and later in code initialize the list if needed

fruch commented on sdcm/cluster_aws.py:
```diff
@@ -85,7 +85,7 @@ def __init__(self, ec2_ami_id, ec2_subnet_id, ec2_security_group_ids,  # pylint:
                  cluster_prefix='cluster',
                  node_prefix='node', n_nodes=10, params=None, node_type=None,
-                 extra_network_interface=False, add_nodes=True):
+                 extra_network_interface=False, add_nodes=True, nodes_smp=[]):
```
don't initialize an argument with a mutable list
set it as None, and later in code initialize the list if needed

fruch commented on sdcm/sct_config.py:
```diff
@@ -499,6 +499,9 @@ class SCTConfiguration(dict):
              In case of random option - Scylla will start with different (random) shards on every node of the cluster
              """),
 
+        dict(name="nodes_smp", env="SCT_NODES_SMP", type=list,
```
```suggestion
        dict(name="nodes_smp", env="SCT_NODES_SMP", type=str_or_list_or_eval,
```

that way one can override it from environment variables if needed

fruch replied:
I wonder maybe we could have a bit more descriptive name for the configuration option
`fixed_smp_per_db_node` 
or
`smp_per_db_node_mapping`

fruch commented on sdcm/sct_config.py:
```diff
@@ -499,6 +499,9 @@ class SCTConfiguration(dict):
              """),
 
+        dict(name="nodes_smp", env="SCT_NODES_SMP", type=list,
+             help="List of shard numbers of nodes in Scylla cluster; list of int, like [4, 5, 3]"),
```
```suggestion
             help="List of shard number to set per node in Scylla cluster; list of int, like [4, 5, 3]"),
```

I wonder how it would work with multi-dc cases:

```yaml
region_name: 'eu-west-1 us-east-1'
n_db_nodes: '2 1'
nodes_smp: [12, 12, 15]
```

Deexie replied:
The number is based on node_index and I think it does not depend on dc

fruch reviewed (dismissed):
LGTM

- we might be able to name a bit better the configuration option
- arguments shouldn't be mutable

Deexie committed:
feature(custom shards longevity): add jenkins pipelines for longevities with custom shard number

Copy asimetric jenkins longevity pipelines and set custom shard
number for them.

Deexie commented:
- use None as a default param value 
- rename nodes_smp to smp_per_db_node_mapping
- use str_or_list_or_eval type for smp_per_db_node_mapping
- add pipelines with custom shard number for some tests that run with random shard num

Deexie commented:
- modify smp_per_db_node_mapping description

Deexie commented:
- drop excessive self arg

denesb commented:
@Deexie what is the status of this?

Deexie commented:
- rebase

fruch commented:
@Deexie what the story with this one ? still needed ?

roydahan commented:
@denesb / @Deexie I think it worth having this functionality and even some of the jobs (or all jobs) that were added here.
This PR introduced a new way to provision cluster with different amount of shards in a controlled way.

denesb commented:
> @denesb / @Deexie I think it worth having this functionality and even some of the jobs (or all jobs) that were added here. This PR introduced a new way to provision cluster with different amount of shards in a controlled way.

Feel free to take over the PR. It would be best if this was worked on by someone who actually plans to use the new infrastructure.
